tracing/src/main/java/com/palantir/tracing/Trace.java

```diff
@@ -30,24 +30,35 @@
  */
 public final class Trace {
 
+    private static final Trace NOOP = new Trace(false, "noop");
+
     private final Deque<OpenSpan> stack;
     private final boolean isObservable;
     private final String traceId;
 
-    private Trace(ArrayDeque<OpenSpan> stack, boolean isObservable, String traceId) {
+    private Trace(Deque<OpenSpan> stack, boolean isObservable, String traceId) {
         checkArgument(!traceId.isEmpty(), "traceId must be non-empty");
 
         this.stack = stack;
         this.isObservable = isObservable;
         this.traceId = traceId;
     }
 
+    static Trace create(boolean isObservable, CharSequence traceId) {
+        if (isObservable) {
+            return new Trace(isObservable, traceId.toString());
+        }
+        return NOOP;
+    }
```

Comment:
I think we still need to generate random trace IDs for non-observable traces, since the ID is applied to requests and logging MDC allowing us to load all the logs for an operation across services, even when the operation has not been sampled.

Fortunately traces are relatively uncommon, spans are created significantly more frequently, and the span ids are unimportant for unsampled traces.

Reply:
Mmm good point, setTrace will always call `MDC.put(Tracers.TRACE_ID_KEY, trace.getTraceId());`, and we really do want those traceIds so we can correlate logs properly.

Reply:
yeah, probably need to remove the `lazyTraceId` then as there isn't a benefit if we're immediately going to generate the `String` form for MDC.

Reply:
It may still be helpful for span IDs, which iirc use the same function. If we have 500 spans per trace, that may provide a sizable improvement.